File: app/src/main/java/br/com/disapps/homepet/ui/login/LoginFragment.java
package br.com.disapps.homepet.ui.login;

import android.content.Intent;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.View;
import android.widget.EditText;

import br.com.disapps.homepet.R;
import br.com.disapps.homepet.app.HomePet;
import br.com.disapps.homepet.ui.common.AppFragment;
import br.com.disapps.homepet.ui.custom.LoadingView;
import br.com.disapps.homepet.ui.profile.ProfileFragment;
import butterknife.BindView;
import butterknife.OnClick;

/**
 * Created by diefferson.santos on 31/08/17.
 */

public class LoginFragment extends AppFragment<ILoginView, LoginPresenter> implements ILoginView {

    @BindView(R.id.username)
    EditText username;

    @BindView(R.id.password)
    EditText password;

    private LoadingView loadingView;

    public static LoginFragment newInstance(){
        return new LoginFragment();
    }

    @Override
    public String getFragmentTag() {
        return LoginFragment.class.getSimpleName();
    }

    @Override
    public int getFragmentLayout() {
        return R.layout.fragment_login;
    }

    @Override
    public LoginPresenter createPresenter() {
        return new LoginPresenter(HomePet.Companion.getInstance().getRestApi(), HomePet.Companion.getInstance().getPreferences());
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getAppActivityListener().setupLoadingActivity(new LoadingView());

        getAppActivityListener().setTitle("Entrar");
    }

    @Override
    public void showLoading(boolean cancelable) {
        if (loadingView != null) {
            loadingView.cancelableOnBackPressed(cancelable);
            loadingView.show(getFragmentManager(), loadingView.getViewTag());
        }
    }

    @Override
    public void dismissLoading() {
        if (loadingView != null) {
            loadingView.dismiss();
        }
    }

    @OnClick(R.id.login_bt)
    public void login(){
        if (validate()){
            getPresenter().login(username.getText().toString(), password.getText().toString());
        }
    }

    @Override
    public void loginSuccessfull(){
        getAppActivityListener().replaceFragment(ProfileFragment.newInstance());
    }

    private boolean validate() {
        boolean valid = true;

        if (TextUtils.isEmpty(username.getText().toString())) {
            valid = false;
            username.setError("Campo obrigatorio");
        }

        if (TextUtils.isEmpty(password.getText().toString())) {
            valid = false;
            password.setError("Campo obrigatorio");
        }

        return valid;
    }
}

File: app/src/main/java/br/com/disapps/homepet/ui/login/LoginFragment.kt
package br.com.disapps.homepet.ui.login

import android.os.Bundle

import android.text.TextUtils
import android.view.View
import br.com.disapps.homepet.R
import br.com.disapps.homepet.app.HomePet
import br.com.disapps.homepet.ui.common.AppFragment
import br.com.disapps.homepet.ui.profile.ProfileFragment
import br.com.disapps.homepet.ui.signup.SignupFragment
import butterknife.OnClick
import kotlinx.android.synthetic.main.fragment_login.*

/**
 * Created by diefferson.santos on 31/08/17.
 */

class LoginFragment : AppFragment<ILoginView, LoginPresenter>(), ILoginView {

    companion object {

        fun newInstance(): LoginFragment {
            return LoginFragment()
        }
    }

    override val fragmentTag: String
        get() = LoginFragment::class.java.simpleName

    override val fragmentLayout: Int
        get() = R.layout.fragment_login

    override fun createPresenter(): LoginPresenter {
        return LoginPresenter(HomePet.instance!!.restApi!!, HomePet.instance!!.preferences!!)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        appActivityListener!!.setTitle("Entrar")
        setupLoadingFragment(loading_view)

        login_bt.setOnClickListener { login() }
        signup_bt.setOnClickListener { signup() }
    }

    private fun login() {
        if (validate()) {
            getPresenter().login(email.text.toString(), password.text.toString())
        }
    }

    private fun signup() {
        appActivityListener!!.replaceAndBackStackFragment(SignupFragment.newInstance())
    }

    override fun loginSuccessfull() {
        appActivityListener!!.replaceFragment(ProfileFragment.newInstance())
    }

    private fun validate(): Boolean {

        var valid = true

        if (TextUtils.isEmpty(email.text.toString())) {
            valid = false
            email.error = "Campo obrigatorio"
        }

        if (TextUtils.isEmpty(password!!.text.toString())) {
            valid = false
            password.error = "Campo obrigatorio"
        }

        return valid
    }


}
